File: distributed-exceptions/src/main/java/io/axoniq/distributedexceptions/command/GiftCard.java
package io.axoniq.distributedexceptions.command;

import io.axoniq.distributedexceptions.api.IssueCardCommand;
import io.axoniq.distributedexceptions.api.CardIssuedEvent;
import io.axoniq.distributedexceptions.api.RedeemCardCommand;
import io.axoniq.distributedexceptions.api.CardRedeemedEvent;
import org.axonframework.eventsourcing.annotation.EventSourcingHandler;
import org.axonframework.eventsourcing.annotation.reflection.EntityCreator;
import org.axonframework.extension.spring.stereotype.EventSourced;
import org.axonframework.messaging.commandhandling.annotation.CommandHandler;
import org.axonframework.messaging.eventhandling.gateway.EventAppender;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.context.annotation.Profile;

import java.lang.invoke.MethodHandles;

@EventSourced(tagKey = "GiftCard")
@Profile("command")
class GiftCard {

    private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());

    private String giftCardId;
    private int remainingValue;

    @CommandHandler
    public static void handle(IssueCardCommand command, EventAppender eventAppender) {
        logger.debug("handling {}", command);
        if (command.amount() <= 0) {
            throw new NegativeOrZeroAmount(command.amount(), "amount <= 0");
        }
        eventAppender.append(new CardIssuedEvent(command.id(), command.amount()));
    }

    @CommandHandler
    public void handle(RedeemCardCommand command, EventAppender eventAppender) {
        logger.debug("handling {}", command);
        if (command.amount() <= 0) {
            throw new NegativeOrZeroAmount(command.amount(), "amount <= 0");
        }
        if (command.amount() > remainingValue) {
            throw new InsufficientFunds("amount > remaining value");
        }
        eventAppender.append(new CardRedeemedEvent(giftCardId, command.amount()));
    }

    @EventSourcingHandler
    public void on(CardIssuedEvent event) {
        logger.debug("applying {}", event);
        giftCardId = event.id();
        remainingValue = event.amount();
        logger.debug("new remaining value: {}", remainingValue);
    }

    @EventSourcingHandler
    public void on(CardRedeemedEvent event) {
        logger.debug("applying {}", event);
        remainingValue -= event.amount();
        logger.debug("new remaining value: {}", remainingValue);
    }

    @EntityCreator
    public GiftCard() {
        // Required by Axon
        logger.debug("Empty constructor invoked");
    }
}

File: distributed-exceptions/src/main/kotlin/io/axoniq/distributedexceptions/command/GiftCard.kt
package io.axoniq.distributedexceptions.command

import io.axoniq.distributedexceptions.api.*
import org.axonframework.commandhandling.CommandHandler
import org.axonframework.eventsourcing.EventSourcingHandler
import org.axonframework.modelling.command.AggregateIdentifier
import org.axonframework.modelling.command.AggregateLifecycle
import org.axonframework.spring.stereotype.Aggregate
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Profile
import java.lang.invoke.MethodHandles

/**
 * @author Stefan Andjelkovic
 */
@Aggregate
@Profile("command")
class GiftCard {
    @AggregateIdentifier
    private var id: String? = null
    private var remainingValue = 0

    constructor() {
        log.debug("empty constructor invoked")
    }

    @CommandHandler
    constructor(cmd: IssueCmd) {
        log.debug("handling {}", cmd)
        requirePositiveAmount(cmd.amount)
        AggregateLifecycle.apply(IssuedEvt(cmd.id, cmd.amount))
    }

    @CommandHandler
    fun handle(cmd: RedeemCmd) {
        log.debug("handling {}", cmd)
        requirePositiveAmount(cmd.amount)
        checkSufficientFunds(cmd)
        AggregateLifecycle.apply(RedeemedEvt(id!!, cmd.amount))
    }

    private fun checkSufficientFunds(cmd: RedeemCmd) {
        if (cmd.amount > remainingValue) {
            throw InsufficientFunds()
        }
    }

    private fun requirePositiveAmount(amount: Int) {
        if (amount <= 0) {
            throw NegativeOrZeroAmount(amount)
        }
    }

    @CommandHandler
    fun handle(cmd: CancelCmd) {
        log.debug("handling {}", cmd)
        AggregateLifecycle.apply(CancelEvt(id!!))
    }

    @EventSourcingHandler
    fun on(evt: IssuedEvt) {
        log.debug("applying {}", evt)
        id = evt.id
        remainingValue = evt.amount
        log.debug("new remaining value: {}", remainingValue)
    }

    @EventSourcingHandler
    fun on(evt: RedeemedEvt) {
        log.debug("applying {}", evt)
        remainingValue -= evt.amount
        log.debug("new remaining value: {}", remainingValue)
    }

    @EventSourcingHandler
    fun on(evt: CancelEvt) {
        log.debug("applying {}", evt)
        remainingValue = 0
        log.debug("new remaining value: {}", remainingValue)
    }

    companion object {
        private val log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass())
    }
}
